packages/expo-router/src/rsc/router/host.tsx

```diff
@@ -63,10 +64,33 @@ if (BASE_PATH === '/') {
 }
 
 if (process.env.EXPO_OS !== 'web' && !window.location?.href) {
+  // This will require a rebuild in bare-workflow to update.
+  const manifest = Constants.expoConfig;
+
+  const originFromConstants =
+    manifest?.extra?.router?.origin ?? manifest?.extra?.router?.generatedOrigin;
+
+  // In legacy cases, this can be extraneously set to false since it was the default before we had a production hosting solution for native servers.
+  if (originFromConstants === false) {
+    const isExpoGo = typeof expo !== 'undefined' && globalThis.expo?.modules?.ExpoGo;
+
+    if (isExpoGo) {
+      // Updating is a bit easier in Expo Go as you don't need a native rebuild.
+      throw new Error(
+        'The "origin" property in the Expo config (app.json) cannot be false when React Server Components is enabled. https://docs.expo.dev/guides/server-components/'
+      );
+    }
+
+    // Add more context about updating the app.json in development builds.
+    throw new Error(
+      'The "origin" property in the Expo config (app.json) cannot be "false" when React Server Components is enabled. Remove the "origin" property from your Expo config and rebuild the native app to resolve. https://docs.expo.dev/guides/server-components/'
```

Comment:
```suggestion
      'The "origin" property in the app config (app.json) cannot be "false" when React Server Components is enabled. Remove the "origin" property from your Expo config and rebuild the native app to resolve. https://docs.expo.dev/guides/server-components/'
```

If my previous comment is valid.